cr-examples/samples/src/main/java/oracle/code/samples/DialectSample.java

```diff
@@ -0,0 +1,281 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package oracle.code.samples;
+
+import jdk.incubator.code.CodeElement;
+import jdk.incubator.code.CodeReflection;
+import jdk.incubator.code.CopyContext;
+import jdk.incubator.code.Op;
+import jdk.incubator.code.OpTransformer;
+import jdk.incubator.code.TypeElement;
+import jdk.incubator.code.Value;
+import jdk.incubator.code.analysis.SSA;
+import jdk.incubator.code.dialect.core.CoreOp;
+import jdk.incubator.code.dialect.java.JavaOp;
+import jdk.incubator.code.dialect.java.JavaType;
+import jdk.incubator.code.interpreter.Interpreter;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Simple example demonstrating how to create new nodes to build a new dialect.
+ * <p>
+ * This example customizes two operations: addition and invoke. Specifically, it
+ * replaces the standard {@link JavaOp.AddOp} node with a custom node
+ * {@link MyAdd}, and the {@link JavaOp.InvokeOp} node with {@link MyInvoke}.
+ * </p>
+ * <p>
+ * After constructing the new code-reflection tree, you can traverse it and
+ * perform new operations with the custom nodes. One potential use case is to
+ * assign specific semantics to these new nodes and handle them accordingly.
+ * </p>
+ *
+ * <p>
+ *     How to run from the terminal?
+ *     <code>
+ *         $ java --enable-preview -cp target/crsamples-1.0-SNAPSHOT.jar oracle.code.samples.DialectSample
+ *     </code>
+ * </p>
+ */
+public class DialectSample {
+
+    // Part A: Create a new Op for specializing additions
+    @CodeReflection
+    public static int mySum(int a, int b) {
+        return a + b;
+    }
+
+    // Custom Node inherits from Op
+    private static class MyAdd extends Op {
+
+        private TypeElement typeElement;
+
+        MyAdd(String name, List<Value> operands, TypeElement typeElement) {
```

Comment:
The operation name should not generally be something the user provides when constructing an operation. The name is for human readability in text, and the key that allows (re-)construction when parsing text back into code models. It's something that should be derived from other inputs and/or constant data. Further, its not something we should really use for pattern matching.

I would like to see if we could remove the name from the state of Op. `OpWriter` of course needs it. Maybe by default the written operation name is the operation (fully qualified) class name, and implementations can override it maybe via a separate interface. We might be able to group the name and externalization of state under one concept for reading/parsing.